File: packages/python/pyfora/NamedSingletons.py
#   Copyright 2015 Ufora Inc.
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.

"""NamedSingletons

A manifest of python objects that have explicit FORA implementations in the purePython module.
"""

pythonSingletonToName = {
    AssertionError: 'AssertionError',
    bool: 'bool',
    dict: 'dict',
    float: 'float',
    int: 'int',
    isinstance: 'isinstance',
    issubclass: 'issubclass',
    list: 'list',
    object: 'object',
    slice: 'slice',
    str: 'str',
    tuple: 'tuple',
    type: 'type',
    type(None): 'NoneType',
    AttributeError: 'AttributeError',
    Exception: 'Exception',
    IndexError: 'IndexError',
    TypeError: 'TypeError',
    NotImplementedError: 'NotImplementedError',
    UserWarning: 'UserWarning',
    UnboundLocalError: 'UnboundLocalError',
    ValueError: 'ValueError',
    ZeroDivisionError: 'ZeroDivisionError'
    }

singletonNameToObject = {v:k for k,v in pythonSingletonToName.iteritems()}

pythonNameToPyforaName = {
    'AssertionError': 'AssertionError',
    'bool': 'BoolType',
    'dict': 'DictType',
    'float': 'FloatType',
    'int': 'IntType',
    'isinstance': 'IsInstance',
    'issubclass': 'IsSubclass',
    'list': 'ListType',
    'object': 'Object',
    'slice': 'SliceType',
    'str': 'StrType',
    'tuple': 'TupleType',
    'type': 'Type',
    'AttributeError': 'AttributeError',
    'Exception': 'Exception',
    'IndexError': 'IndexError',
    'NoneType': 'NoneType',
    'NotImplementedError': 'NotImplementedError',
    'TypeError': 'TypeError',
    'UserWarning': 'UserWarning',
    'UnboundLocalError': 'UnboundLocalError',
    'ValueError': 'ValueError',
    'ZeroDivisionError': 'ZeroDivisionError'
    }

File: packages/python/pyfora/src/NamedSingletons.cpp
/***************************************************************************
   Copyright 2016 Ufora Inc.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
****************************************************************************/
#include "NamedSingletons.hpp"
#include "PyObjectUtils.hpp"

#include <stdexcept>


NamedSingletons::NamedSingletons()
    : mSingletonNameToObjectDict(nullptr)
    {
    PyObject* namedSingletonsModule = PyImport_ImportModule("pyfora.NamedSingletons");
    if (namedSingletonsModule == nullptr) {
        throw std::runtime_error(
            "error getting pyfora.NamedSingletons: " +
            PyObjectUtils::exc_string()
            );
        }

    mSingletonNameToObjectDict = PyObject_GetAttrString(
        namedSingletonsModule,
        "singletonNameToObject");
    if (mSingletonNameToObjectDict == nullptr) {
        throw std::runtime_error(
            "error getting pyfora.NamedSingletons.singletonNameToObject: " +
            PyObjectUtils::exc_string()
            );
        }
    if (not PyDict_Check(mSingletonNameToObjectDict)) {
        Py_DECREF(mSingletonNameToObjectDict);
        throw std::runtime_error(
            "expected pyfora.NamedSingletons.singletonNameToObject "
            " to be a dict"
            );
        }
    }


PyObject* NamedSingletons::singletonNameToObject(const std::string& s)
    {
    PyObject* tr = PyDict_GetItemString(
        _getInstance().mSingletonNameToObjectDict,
        s.c_str()
        );

    // borrowed reference, so we incref it
    Py_INCREF(tr);

    return tr;
    }
